locked-asset/simple-lock-energy/src/penalty.rs

```diff
@@ -0,0 +1,80 @@
+use common_structs::Epoch;
+
+use crate::lock_options::EPOCHS_PER_YEAR;
+
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+#[derive(TypeAbi, TopEncode, TopDecode)]
+pub struct PenaltyPercentage {
+    pub first_threshold: u16,
+    pub second_threshold: u16,
+    pub third_threshold: u16,
+}
+
+#[elrond_wasm::module]
+pub trait LocalPenaltyModule {
+    fn calculate_penalty_percentage_full_unlock(
+        &self,
+        lock_epochs_remaining: Epoch,
+        penalty_percentage_struct: &PenaltyPercentage,
+    ) -> u64 {
+        let first_threshold_penalty = penalty_percentage_struct.first_threshold as u64;
+        let second_threshold_penalty = penalty_percentage_struct.second_threshold as u64;
+        let third_threshold_penalty = penalty_percentage_struct.third_threshold as u64;
+
+        match lock_epochs_remaining / (EPOCHS_PER_YEAR + 1u64) {
+            0 => first_threshold_penalty * lock_epochs_remaining / EPOCHS_PER_YEAR,
+            1 => {
+                // value between 0 and 360
+                let normalized_current_epoch_unlock = lock_epochs_remaining - EPOCHS_PER_YEAR;
+                first_threshold_penalty
+                    + (second_threshold_penalty - first_threshold_penalty)
+                        * normalized_current_epoch_unlock
+                        / EPOCHS_PER_YEAR
+            }
+            2 | 3 => {
+                // value between 721 and 1440 epochs (years 3,4) normalized to 0 - 720
+                let normalized_current_epoch_unlock = lock_epochs_remaining - (2 * EPOCHS_PER_YEAR);
+                second_threshold_penalty
+                    + (third_threshold_penalty - second_threshold_penalty)
+                        * normalized_current_epoch_unlock
+                        / (2 * EPOCHS_PER_YEAR)
+            }
+            _ => sc_panic!("Invalid unlock choice"),
+        }
+    }
+
+    fn calculate_epoch_from_penalty_percentage(self, penalty_percentage: u64) -> Epoch {
+        let penalty_percentage_struct = self.penalty_percentage().get();
```

Comment:
You could pass this as a parameter as well, like in the function from above, as you already have this variable available in the calling function. That way, you won't read from the storage twice.